test/hiccup/compiler_test.clj

```diff
@@ -0,0 +1,12 @@
+(ns hiccup.compiler-test
+  (:require [clojure.test :refer :all]
+            [hiccup.compiler :as compiler]))
+
+(deftest test-concatenate-strings
+  (let [concatenate-strings #'compiler/concatenate-strings]
+    (testing "empty collection"
+      (is (= [] (concatenate-strings nil)))
+      (is (= [] (concatenate-strings []))))
+    (testing "concatenates consecutive strings, ignores other types"
+      (is (= ["abc"] (concatenate-strings ["a" "b" "c"])))
+      (is (= [1 "ab" 2 "cd" 3] (concatenate-strings [1 "a" "b" 2 "c" "d" 3]))))))
```

Comment:
We shouldn't be testing private functions on their own, as it should be possible to refactor the internals of a namespace without causing the tests to fail.

Can we perform the same effective test in `hiccup.core_test`, using `html`?

Reply:
It's only used by the macro, so an integration test would not be able check whether `concatenate-strings` works or whether it's just an alias for `identity`.

Reply:
But we can check the generated code with `macroexpand-1`, right?

Reply:
Checking the outcome of macroexpand would produce extremely fragile tests. It would hinder refactoring instead of helping it, because it would cause refactoring to break tests.

Reply:
We could do it that way, but it seems simpler just to count the forms and use that as an approximation for the size of the generated bytecode.

To put it another way, it's unlikely that the bytecode is going to significantly increase in size if the number of forms remains the same. The point of tests is to alert us of future regressions, after all. We just need a sufficiently good canary in our coalmine; it doesn't need to be perfect.

Reply:
Yes, that seems like a good compromise. Counting forms is easier.

Reply:
I've now written such tests.

Reply:
Thanks! The new tests look good. Can you remove this compiler test and namespace from the PR?